File: tests/cpu/seats/home-continent-rows.test.ts
import { describe, it, expect } from 'vitest';
import { makeMap, makeState, tileAtCoords, settleAt } from '../helpers';
import { emptySeat, seatOf, onHomeContinent } from '../../../cpu/core/seats';
import { getModifiers, prodMultFor } from '../../../cpu/core/effects';
import { rosterRouteCapacity } from '../../../cpu/core/trade';
import { PROD_MULT_ROWS, ROUTE_CAPACITY_ROWS } from '../../../cpu/data/civilizations';
import { CIV_LEADERS } from '../../../cpu/data/seats';
import type { GameState } from '../../../cpu/core/types';

/**
 * The three clauses that read a city's landmass against its seat's HOME one
 * (C-48): Spain's district Production off the capital's continent, Victoria's
 * Trade Route capacity per foreign-continent city, and Phoenicia's 100%-loyal
 * coastal cities at home.
 *
 * The GPU twin is tests/gpu/home_continent_rows_test.py.
 */
const civRow = (civ: string) => CIV_LEADERS.findIndex((l) => l.civ === civ);
const leadRow = (leader: string) => CIV_LEADERS.findIndex((l) => l.leader === leader);

/** Two landmasses split by a column of ocean; column 12 is the sea. */
function scene(row: number): GameState {
  const map = makeMap(24, 24, 'GRASSLAND');
  for (const t of map.tiles) if (t.col === 12) t.terrain = 'OCEAN';
  const state = makeState(map);
  state.seats.push(emptySeat(1));
  state.seats[0].civ = row;
  return state;
}

describe("Spain's cities off the capital's continent", () => {
  it('reads the install: +25% toward every district item', () => {
    const rows = PROD_MULT_ROWS.filter((r) => r.offHomeContinent);
    expect(rows.length).toBe(1);
    expect(rows[0].civ).toBe('SPAIN');
    expect(rows[0].pct).toBe(25);
    expect(rows[0].every).toBe('district');
  });

  it('multiplies a district only in a city off the home continent', () => {
    const state = scene(civRow('SPAIN'));
    settleAt(state, tileAtCoords(state.map, 4, 5).index, 0);
    const rows = getModifiers(state, 0).prodMults;
    const item = { kind: 'district' as const, districtItem: 'CAMPUS' };
    expect(prodMultFor(rows, item, true)).toBeCloseTo(1.25);
    expect(prodMultFor(rows, item, false)).toBe(1);
    // ...and it is a DISTRICT clause, not a blanket one
    expect(prodMultFor(rows, { kind: 'building', building: 'MONUMENT' }, true)).toBe(1);
  });

  it('pays a seat the roster does not name nothing either way', () => {
    const state = scene(-1);
    settleAt(state, tileAtCoords(state.map, 4, 5).index, 0);
    const rows = getModifiers(state, 0).prodMults;
    expect(prodMultFor(rows, { kind: 'district', districtItem: 'CAMPUS' }, true)).toBe(1);
  });
});

describe("Victoria's capacity per foreign-continent city", () => {
  it('reads the install: one per such city', () => {
    const rows = ROUTE_CAPACITY_ROWS.filter((r) => r.perForeignCity);
    expect(rows.length).toBe(1);
    expect(rows[0].leader).toBe('VICTORIA');
    expect(rows[0].amount).toBe(1);
  });

  it('counts the cities abroad and nothing at home', () => {
    const state = scene(leadRow('VICTORIA'));
    settleAt(state, tileAtCoords(state.map, 4, 5).index, 0);
    expect(rosterRouteCapacity(state, 0)).toBe(0);
    settleAt(state, tileAtCoords(state.map, 4, 12).index, 0);   // same landmass
    expect(rosterRouteCapacity(state, 0)).toBe(0);
    settleAt(state, tileAtCoords(state.map, 18, 5).index, 0);   // across the water
    expect(rosterRouteCapacity(state, 0)).toBe(1);
    settleAt(state, tileAtCoords(state.map, 18, 12).index, 0);  // and another
    expect(rosterRouteCapacity(state, 0)).toBe(2);
  });
});

describe("Phoenicia's coastal cities at home", () => {
  it('is 100% loyal only when coastal AND on the home continent', () => {
    const state = scene(civRow('PHOENICIA'));
    const m = getModifiers(state, 0);
    expect(m.coastalHomeLoyal).toBe(true);
    // the capital's landmass is the home one by construction
    const home = settleAt(state, tileAtCoords(state.map, 11, 5).index, 0);  // beside the sea
    expect(onHomeContinent(state, 0, home.centerIndex)).toBe(true);
    // a city across the water is NOT on the home continent, so the clause
    // cannot reach it however coastal it is
    const abroad = settleAt(state, tileAtCoords(state.map, 13, 15).index, 0);
    expect(onHomeContinent(state, 0, abroad.centerIndex)).toBe(false);
  });

  it('names Phoenicia alone', () => {
    for (const civ of ['SPAIN', 'ENGLAND', 'ROME']) {
      const state = scene(civRow(civ));
      expect(getModifiers(state, 0).coastalHomeLoyal, civ).toBe(false);
    }
    expect(seatOf(scene(-1), 0)!.civ).toBe(-1);
  });
});

File: tests/gpu/home_continent_rows_test.py
"""THE HOME-CONTINENT ROWS — the GPU half.

    npm run seed && npm run export        # (once) writes seeder/worlds/
    python tests/gpu/home_continent_rows_test.py

The TS twin is tests/cpu/seats/home-continent-rows.test.ts.

The three clauses that read a city's landmass against its seat's HOME one
(C-48): Spain's district Production off the capital's continent, Victoria's
Trade Route capacity per foreign-continent city, and Phoenicia's 100%-loyal
coastal cities at home. No fixture seats any of the three, so these lanes are
the only evidence any of them fires.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

import torch

sys.path.insert(0, str(Path(__file__).resolve().parent.parent.parent / "gpu"))
sys.path.insert(0, str(Path(__file__).resolve().parent))
from core import BatchSim, load_rules, load_fixture, fixture_paths
from warmup import settle_all

B0 = 0
ROW = 1


def build(path) -> BatchSim:
    return settle_all(BatchSim([load_fixture(path)], load_rules(),
                               device="cpu", dtype=torch.float64))


def _two_continent_fixture():
    for p in fixture_paths():
        fx = json.loads(Path(p).read_text(encoding="utf-8"))
        if len({int(t.get("cont", -1)) for t in fx["tiles"]} - {-1}) >= 2:
            return p
    raise AssertionError("no fixture carries two landmasses")


def _seat(sim, row: int, civ=None, leader=None) -> None:
    if civ is None and leader is None:
        sim.row_civ[B0, row] = -1
        sim.row_leader[B0, row] = -1
    elif leader is not None:
        li = sim._leader_idx(leader)
        sim.row_civ[B0, row] = sim._pair_civ[li]
        sim.row_leader[B0, row] = li
    else:
        ci = sim._civ_ids.index(civ)
        sim.row_civ[B0, row] = ci
        sim.row_leader[B0, row] = sim._pair_civ.index(ci)
    sim._eff_version += 1
    sim._gen_ver += 1
    sim._bldg_version += 1


CITY_MIN_DIST = 4    # world/types.ts — every centre keeps this much room


def _found_abroad(sim, row: int) -> int:
    """Found one city on a landmass that is NOT the row's home; answer its tile.
    The site must clear CITY_MIN_DIST of every existing centre, own or not, or
    the founding is refused and the lane measures nothing."""
    home = int(sim._home_continent(row)[B0])
    assert home >= 0
    centres = [int(t) for t in sim.centre_slot_at[B0].nonzero().flatten().tolist()
               if int(sim.centre_slot_at[B0, t]) >= 0]
    site = None
    for t in range(sim.T):
        if int(sim.tile_continent[B0, t]) < 0 or int(sim.tile_continent[B0, t]) == home:
            continue
        if int(sim.tile_seat[B0, t]) >= 0 or bool(sim.water[B0, t]):
            continue
        if not bool(sim.passable[B0, t]) or int(sim.district[B0, t]) >= 0:
            continue
        if any(int(sim.pair_dist[t, c]) < CITY_MIN_DIST for c in centres):
            continue
        site = t
        break
    assert site is not None, "no free, well-spaced tile on another landmass"
    sim._found_city_at(row, torch.tensor([True]), torch.tensor([site]))
    assert int(sim.tile_seat[B0, site]) == row, "the founding did not land"
    return site


def test_the_wire(rules, path) -> None:
    sim = build(path)
    prod = [r for r in sim._prod_mult_rows if r[9]]
    assert len(prod) == 1, f"one off-home production row expected, wire has {len(prod)}"
    assert float(prod[0][5]) == 25.0 and prod[0][7] == 3, "not +25% on EVERY district item"
    cap = [r for r in sim._route_cap_rows if r[7]]
    assert len(cap) == 1 and cap[0][2] == 1, "one capacity row of amount 1 expected"
    assert bool(sim._coastal_home_loyal.any()), "no civilization is coastal-home loyal"
    assert int(sim._coastal_home_loyal.long().sum()) == 1, "more than one claims the clause"
    print("  1 the wire OK — +25% districts, +1 capacity per city, one loyal civ")


def test_spain_production_is_off_home_only(rules, path) -> None:
    sim = build(path)
    _seat(sim, ROW, civ="SPAIN")
    off = ~sim._on_home_continent(ROW, sim.city_center[:, ROW])
    # the row's own gate, read exactly as the queue reads it
    assert not bool(off[B0, 0]), "the FIRST city is off its own home continent"
    site = _found_abroad(sim, ROW)
    off2 = ~sim._on_home_continent(ROW, sim.city_center[:, ROW])
    col = next(j for j in range(sim.RC)
               if bool(sim.city_alive[B0, ROW, j]) and int(sim.city_center[B0, ROW, j]) == site)
    assert bool(off2[B0, col]), "the city founded abroad did not read as off-home"
    assert not bool(off2[B0, 0]), "the capital's own city drifted off-home"
    # ...and RUN the production path at B > 1, which is where the row is
    # actually read. The predicate alone missed a shape fault here, and so
    # did a B=1 run: a full-width [B, RC] read BROADCASTS silently against a
    # [B] one when B is 1, and only reds when the batch is wider (the
    # battery's shard was B=3).
    wide = settle_all(BatchSim([load_fixture(path), load_fixture(path)],
                               load_rules(), device="cpu", dtype=torch.float64))
    assert wide.B > 1, "this lane needs a batch wider than one to mean anything"
    _seat(wide, ROW, civ="SPAIN")
    # a DISTRICT must actually be at the queue head, or the row's own arm
    # never runs and the lane passes without touching it
    for j in (0, 1):
        wide.city_current[:, ROW, j, 0] = wide.DISTRICT_BASE
        wide.city_qtile[:, ROW, j, 0] = wide.city_center[:, ROW, j]
    wide._eff_version += 1
    for j in (0, 1):
        jc = torch.full((wide.B,), j, dtype=torch.long)
        wide._seat_city_produce(ROW, jc, torch.ones(wide.B, dtype=torch.bool),
                                torch.full((wide.B,), 10.0, dtype=wide.dtype))
    print("  2 Spain OK — the gate reads right, and the production path runs",
          "at B =", wide.B)


def test_victoria_capacity_counts_cities_abroad(rules, path) -> None:
    sim = build(path)
    _seat(sim, ROW, leader="VICTORIA")
    base = int(sim._roster_route_capacity(ROW)[B0])
    assert base == 0, f"a seat with every city at home already had {base} capacity"
    _found_abroad(sim, ROW)
    one = int(sim._roster_route_capacity(ROW)[B0])
    assert one == 1, f"one city abroad gave {one} capacity, expected 1"
    _found_abroad(sim, ROW)
    two = int(sim._roster_route_capacity(ROW)[B0])
    assert two == 2, f"two cities abroad gave {two} capacity, expected 2"
    # ...and a seat the roster does not name counts nothing
    _seat(sim, ROW, None)
    assert int(sim._roster_route_capacity(ROW)[B0]) == 0, "a plain seat was paid"
    print("  3 Victoria OK — one per city abroad, nothing for a plain seat")


def test_phoenicia_is_loyal_coastal_at_home(rules, path) -> None:
    sim = build(path)
    _seat(sim, ROW, civ="PHOENICIA")
    ci = int(sim.row_civ[B0, ROW])
    assert bool(sim._coastal_home_loyal[ci]), "Phoenicia does not carry the clause"
    # every other civilization the roster holds does NOT
    others = [c for c in range(len(sim._civ_ids)) if c != ci and bool(sim._coastal_home_loyal[c])]
    assert not others, f"other civilizations claim the clause: {others}"
    # the clause needs BOTH halves: coastal, and the home landmass
    ctr = sim.city_center[:, ROW, 0]
    assert bool(sim._on_home_continent(ROW, ctr)[B0]), "the capital is not on its own landmass"
    site = _found_abroad(sim, ROW)
    away = torch.full((sim.B,), site, dtype=torch.long)
    assert not bool(sim._on_home_continent(ROW, away)[B0]), \
        "a city across the water read as the home continent"
    print("  4 Phoenicia OK — the clause is hers, and needs the home landmass")


def main() -> int:
    rules = load_rules()
    path = _two_continent_fixture()
    test_the_wire(rules, path)
    test_spain_production_is_off_home_only(rules, path)
    test_victoria_capacity_counts_cities_abroad(rules, path)
    test_phoenicia_is_loyal_coastal_at_home(rules, path)
    print("BATTERY OK home_continent_rows")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
